### PR TITLE
Pre and Post rule with KubevirtVm

### PR DESCRIPTION
**Pull request by atrivedi-px**
**What this PR does / why we need it**:
Run kubevirt tests with freeze/thaw rules

**Which issue(s) this PR fixes** (optional)
Closes #PB-4940

**Special notes for your reviewer**:
Aetos Dashboard: https://aetos.pwx.purestorage.com/resultSet/testSetID/441405

### CODE REVIEW

File: drivers/backup/portworx/portworx.go [hunk 1]
```diff
@@ -2094,6 +2096,21 @@ var (
 	}
 )
 
+var (
+	// KubeVirtRule template contains the template for pre, post rules
+	// and pod selector to be used for kubevirt backup
+	KubevirtRuleTemplate = map[string]map[string]string{
```
Comment (kphalgun-px) [inline drivers/backup/portworx/portworx.go:2102]: backup.RuleSpec?

File: tests/backup/backup_kubevirt_test.go [hunk 2]
```diff
@@ -141,8 +142,16 @@ var _ = Describe("{KubevirtVMBackupRestoreWithDifferentStates}", func() {
...
 					defer GinkgoRecover()
 					defer wg.Done()
+					_, preRuleName, err := CreateKubevirtBackupRuleForAllVMsInNamespace(ctx, appCtx.ScheduleOptions.Namespace, "pre", "default")
+					log.FailOnError(err, "Unable to create Pre Rule")
+					preRuleUid, err := Inst().Backup.GetRuleUid(orgID, ctx, preRuleName)
+					log.FailOnError(err, "Unable fetch pre rule uid")
+					_, postRuleName, err := CreateKubevirtBackupRuleForAllVMsInNamespace(ctx, appCtx.ScheduleOptions.Namespace, "post", "default")
+					log.FailOnError(err, "Unable to create Post Rule")
+					postRuleUid, err := Inst().Backup.GetRuleUid(orgID, ctx, postRuleName)
+					log.FailOnError(err, "Unable fetch post rule uid")
```
Comment (kphalgun-px) [inline tests/backup/backup_kubevirt_test.go:152]: include rule name in logs?

**Review by kphalgun-px — APPROVED** [review]
LGTM

File: drivers/backup/portworx/portworx.go (hunk 1 above)
Comment (atrivedi-px) [reply drivers/backup/portworx/portworx.go]: Renamed this to KubevirtRuleSpec, for other applications the rules are saved as AppParameters in same file.

File: tests/backup/backup_kubevirt_test.go (hunk 2 above)
Comment (atrivedi-px) [reply tests/backup/backup_kubevirt_test.go]: Added

File: drivers/backup/portworx/portworx.go (hunk 1 above)
Comment (kphalgun-px) [reply drivers/backup/portworx/portworx.go]: var KubevirtRuleSpec = map[string]KubevirtRule{
	"default": {
		PreRule: PreRule{
			Rule: RuleSpec{
				ActionList:      []string{"/usr/bin/virt-freezer --freeze --name <vm-name> --namespace <namespace>"},
				PodSelectorList: []string{"vm.kubevirt.io/name=<vm-name>"},
				Background:      []string{"false"},
				RunInSinglePod:  []string{"false"},
				Container:       []string{},
			},
		},
		PostRule: PostRule{
			Rule: RuleSpec{
				ActionList:      []string{"/usr/bin/virt-freezer --unfreeze --name <vm-name> --namespace <namespace>"},
				PodSelectorList: []string{"vm.kubevirt.io/name=<vm-name>"},
				Background:      []string{"false"},
				RunInSinglePod:  []string{"false"},
				Container:       []string{},
			},
		},
	},
}

How about something like this? *RunInSinglePod should have been boolean.

Comment (kphalgun-px) [reply drivers/backup/portworx/portworx.go]: As discussed it is better to keep it as is for consistency. Thank you.

File: drivers/backup/portworx/portworx.go [hunk 3]
```diff
@@ -2196,6 +2213,107 @@ func (p *portworx) CreateRuleForBackup(appName string, orgID string, prePostFlag
...
+			Name:  ruleName,
+			OrgId: orgID,
+		},
+		RulesInfo: &rulesInfo,
+	}
+	// ctx, err := backup.GetAdminCtxFromSecret()
+	// if err != nil {
+	// 	err = fmt.Errorf("Failed to fetch px-central-admin ctx: [%v]", err)
+	// 	return false, ruleName, err
+	// }
```
Comment (mkoppal-px) [inline drivers/backup/portworx/portworx.go:2286]: Remove the commented out code

Comment (atrivedi-px) [reply drivers/backup/portworx/portworx.go]: Done

File: tests/backup/backup_kubevirt_test.go [hunk 4]
```diff
@@ -141,8 +142,18 @@ var _ = Describe("{KubevirtVMBackupRestoreWithDifferentStates}", func() {
...
+					preRuleUid, err := Inst().Backup.GetRuleUid(orgID, ctx, preRuleName)
+					log.FailOnError(err, "Unable fetch pre rule uid")
+					_, postRuleName, err := CreateKubevirtBackupRuleForAllVMsInNamespace(ctx, appCtx.ScheduleOptions.Namespace, "post", "default")
+					log.FailOnError(err, "Unable to create Post Rule")
+					log.Infof("Post rule Name - [%s]", postRuleName)
+					postRuleUid, err := Inst().Backup.GetRuleUid(orgID, ctx, postRuleName)
+					log.FailOnError(err, "Unable fetch post rule uid")
 					log.InfoD("creating backup [%s] in source cluster [%s] (%s), organization [%s], of namespace [%s], in backup location [%s]", backupName, SourceClusterName, sourceClusterUid, orgID, appCtx.ScheduleOptions.Namespace, backupLocationName)
-					err := CreateBackupWithValidation(ctx, backupName, SourceClusterName, backupLocationName, backupLocationUID, []*scheduler.Context{appCtx}, labelSelectors, orgID, sourceClusterUid, "", "", "", "")
+					err = CreateBackupWithValidation(ctx, backupName, SourceClusterName, backupLocationName, backupLocationUID, []*scheduler.Context{appCtx}, labelSelectors, orgID, sourceClusterUid, preRuleName, preRuleUid, postRuleName, postRuleUid)
```
Comment (mkoppal-px) [inline tests/backup/backup_kubevirt_test.go:156]: Please apply the pre and post rule to all the backups taken as part of this test case and the Upgrade test case

Comment (atrivedi-px) [reply tests/backup/backup_kubevirt_test.go:156]: Done, dashboard with run - https://aetos.pwx.purestorage.com/resultSet/testSetID/452940

**Pull request merged**